File: app/src/oss/java/james/metronome/billing/OssBillingProvider.java
package james.metronome.billing;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.ImageView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.bumptech.glide.Glide;

import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.core.content.ContextCompat;
import james.metronome.Metronome;
import james.metronome.R;

public class OssBillingProvider implements BillingInterface {

    private static final String URL_DONATE = "https://jfenn.me/links/liberapay";
    private static final String KEY_DONATE = "james.metronome.billing.OssBillingProvider.KEY_DONATE";

    public OssBillingProvider(Metronome metronome) {
        // nope
    }

    @Override
    public void onCreateActivity() {
        // nope
    }

    @Override
    public void onDestroyActivity() {
        // nope
    }

    @Override
    public String getPrice(Context context) {
        return context.getString(R.string.action_donate);
    }

    @Override
    public String getSku(Context context) {
        // nope
        return null;
    }

    @Override
    public boolean isPremium(Context context) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        return prefs.getBoolean(KEY_DONATE, false);
    }

    @Override
    public void onPremium(Activity activity) {
        if (!isPremium(activity)) {
            View view = LayoutInflater.from(activity).inflate(R.layout.dialog_premium, null);
            Glide.with(activity).load("https://jfenn.me/images/headers/metronomePremium.png").into((ImageView) view.findViewById(R.id.image));

            AppCompatCheckBox checkBox = view.findViewById(R.id.hideCheckBox);
            checkBox.setVisibility(View.VISIBLE);
            checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
                SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(activity);
                prefs.edit().putBoolean(KEY_DONATE, isChecked).apply();
            });

            new MaterialDialog.Builder(activity)
                    .customView(view, false)
                    .backgroundColor(Color.WHITE)
                    .cancelable(false)
                    .positiveText(getPrice(activity))
                    .positiveColor(ContextCompat.getColor(activity, R.color.colorAccent))
                    .onPositive((dialog, which) -> {
                        buyPremium(activity);
                        dialog.dismiss();
                    })
                    .negativeText(R.string.title_use_anyway)
                    .negativeColor(ContextCompat.getColor(activity, R.color.textColorSecondaryInverse))
                    .onNegative((dialog, which) -> dialog.dismiss())
                    .show();
        }
    }

    @Override
    public void buyPremium(Activity activity) {
        activity.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(URL_DONATE)));
    }

    @Override
    public void onPremiumBought(int resultCode, Intent data) {
        // nope
    }
}

File: app/src/oss/java/me/jfenn/metronome/billing/OssBillingProvider.kt
package me.jfenn.metronome.billing

import android.content.Intent
import android.net.Uri
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.widget.CompoundButton
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatCheckBox
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.bumptech.glide.Glide
import me.jfenn.metronome.Metronome
import me.jfenn.metronome.R
import me.jfenn.metronome.utils.PreferenceDelegate

class OssBillingProvider(
        private val metronome: Metronome
) : BillingInterface {

    private var hasPremium by PreferenceDelegate(metronome, KEY_DONATE, false)

    override fun onCreateActivity() {
        // nope
    }

    override fun onDestroyActivity() {
        // nope
    }

    override fun getPrice(): String {
        return metronome.getString(R.string.action_donate)
    }

    override fun getSku(): String {
        // nope
        return ""
    }

    override fun isPremium(): Boolean {
        val prefs = PreferenceManager.getDefaultSharedPreferences(metronome)
        return prefs.getBoolean(KEY_DONATE, false)
    }

    override fun onPremium(activity: AppCompatActivity) {
        if (!hasPremium) {
            val view = LayoutInflater.from(activity).inflate(R.layout.dialog_premium, null)

            Glide.with(activity)
                    .load("https://jfenn.me/images/headers/metronomePremium.png")
                    .into(view.findViewById(R.id.image))

            view.findViewById<AppCompatCheckBox>(R.id.hideCheckBox).apply {
                visibility = View.VISIBLE
                setOnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean ->
                    hasPremium = isChecked
                }
            }

            MaterialDialog(activity).show {
                customView(view = view, noVerticalPadding = true)
                cancelable(false)
                positiveButton(text = price) {
                    buyPremium(activity)
                    it.dismiss()
                }
                negativeButton(R.string.title_use_anyway) {
                    it.dismiss()
                }
            }
        }
    }

    override fun buyPremium(activity: AppCompatActivity) {
        activity.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(URL_DONATE)))
    }

    companion object {
        private const val URL_DONATE = "https://jfenn.me/links/liberapay"
        private const val KEY_DONATE = "me.jfenn.metronome.billing.OssBillingProvider.KEY_DONATE"
    }
}